LFS-172: As an admin, I can import, visualize and design questionnaires

nela-jankechova opened the pull request:
In this PR:
- Create and delete questionnaires
- Visualize, create, edit, and delete questions and sections
- Add new field types using JSON files

TODO: 
- Allow conditional statements to be shown and edited
- Allow drag & drop reordering of questions and sections
- Show a column in questionnaires with the number of forms which links to a page displaying the forms

nela-jankechova committed:
LFS-172: Add a dialog for editing a questionnaire

nela-jankechova committed:
LFS-172: Add a new questionnaire dialog

nela-jankechova committed:
LFS-172: Update questionnaires and questionnaire

nela-jankechova committed:
LFS-172: Seperate Edit and Delete Dialogs

nela-jankechova committed:
LFS-172: Delete former Edit Dialog page

nela-jankechova committed:
LFS-172: Incorporate Typography

nela-jankechova committed:
LFS-172: Add Actions to LiveTable

nela-jankechova committed:
LFS-172: Add number of Forms to Delete Dialog

nela-jankechova committed:
LFS-172: Delete Questionnaire Dialog

nela-jankechova committed:
LFS-172: Show Sections in Questionnaire

nela-jankechova committed:
LFS-172: Fix import

nela-jankechova committed:
Create JSON files

nela-jankechova committed:
Refactor "Edit" dialogs to use JSON

nela-jankechova committed:
LFS-172: Update delete and edit questionnaire

nela-jankechova committed:
LFS-172: Update Object Input

nela-jankechova committed:
LFS-172: Enable form submission

nela-jankechova committed:
LFS-172: Create Answer Options component

nela-jankechova committed:
LFS-172: Refactor

nela-jankechova committed:
LFS-172: Comment Input Components

nela-jankechova committed:
LFS-172: Bug Fixes

nela-jankechova committed:
LFS-172: Remove Forms column

nela-jankechova committed:
LFS-172: Remove empty file

sashaandjic commented:
- deleting of a question works well, however it gets deleted once you refresh
- when adding a new question letters "s" and "q" do not work if used as a first character
-  hints do not work
- min/max value does not seem to work

nela-jankechova committed:
Update AnswerOptions.jsx

nela-jankechova committed:
Update EditDialog.jsx

nela-jankechova committed:
Update LongInput.jsx

nela-jankechova committed:
Update Questionnaire.jsx

nela-jankechova committed:
LFS-172: Fix EditDialog

nela-jankechova commented:
The following issues existed previously so I did not address them: 
deleting of a question works well, however it gets deleted once you refresh
hints do not work
min/max value does not seem to work

OrdiNeu commented on modules/data-entry/src/main/frontend/src/questionnaire/Section.json:
```diff
@@ -0,0 +1,8 @@
+    "recurrent": "boolean",
+    "initialNumberOfInstances" : "long"
+  }
+]
```
Every file should have a newline at the end of the file

OrdiNeu commented on modules/data-entry/src/main/frontend/src/questionnaire/Questionnaire.jsx:
```diff
@@ -77,57 +109,206 @@ let Questionnaire = (props) => {
+      { data &&
+        <Grid item>
+          <Button aria-controls="simple-menu" aria-haspopup="true" onClick={handleOpenMenu}>
+            Add...
```
This button should definitely be at the bottom of the form (after the questions), since that is where the new section/question is added.

nela-jankechova replied:
This is based on the diagram which Marta created. I can understand why she would put it at the top, because otherwise if you have many questions you have to scroll to the bottom just to see/access this functionality. Maybe we can discuss it a bit more?

OrdiNeu commented on modules/data-entry/src/main/frontend/src/questionnaire/StringInput.jsx:
```diff
@@ -0,0 +1,40 @@
+};
+
+export default withStyles(QuestionnaireStyle)(StringInput);
+  
```
```suggestion

```
I think the spaces here are preventing GitHub from recognizing the newline

OrdiNeu commented on modules/data-entry/src/main/frontend/src/questionnaire/Question.json:
```diff
@@ -0,0 +1,50 @@
+        }
+      }
+    }
+  ]
```
```suggestion
  ]

```

OrdiNeu commented on modules/data-entry/src/main/frontend/src/questionnaire/ObjectInput.jsx:
```diff
@@ -0,0 +1,76 @@
+  data: PropTypes.object.isRequired,
+};
+  
+export default withStyles(QuestionnaireStyle)(ObjectInput);
```
```suggestion
export default withStyles(QuestionnaireStyle)(ObjectInput);

```

OrdiNeu commented on modules/data-entry/src/main/frontend/src/questionnaire/Fields.jsx:
```diff
@@ -0,0 +1,92 @@
+  JSON: PropTypes.object.isRequired
+};
+  
+export default withStyles(QuestionnaireStyle)(Fields);
```
```suggestion
export default withStyles(QuestionnaireStyle)(Fields);

```

OrdiNeu commented on modules/data-entry/src/main/frontend/src/questionnaire/BooleanInput.jsx:
```diff
@@ -0,0 +1,44 @@
+};
+
+export default withStyles(QuestionnaireStyle)(BooleanInput);
+  
```
```suggestion

```

OrdiNeu commented on modules/data-entry/src/main/frontend/src/questionnaire/DeleteDialog.jsx:
```diff
@@ -0,0 +1,126 @@
+    const left = window.top.outerWidth / 2 + window.top.screenX - ( width / 2);
+    // After a successful log in, the login dialog code will "open" the specified resource, which results in executing the specified javascript code
+    window.open("/login.html?resource=javascript%3Awindow.close()", "loginPopup", `width=${width}, height=${height}, top=${top}, left=${left}`);
+    // Display 'save'ss on button
```
```suggestion
    // Display 'save' on button
```

OrdiNeu commented on modules/data-entry/src/main/frontend/src/questionnaire/Fields.jsx:
```diff
@@ -0,0 +1,92 @@
+
+    let inputTypes = (key, value) => {
+    return [
+      {
+        dataType: 'boolean',
+        score: ( value === 'boolean' ? 60 : 10),
+        component: (
+          <Grid container alignItems='flex-end' spacing={2} key={key}>
+            <Grid item xs={6}><Typography>{ formatString(key) }</Typography></Grid>
+            <Grid item xs={6}><BooleanInput objectKey={key} data={data}/></Grid>
+          </Grid>
+        )
+      },
+      {
+        dataType: 'string',
+        score: ( value === 'string' ? 60 : 40),
+        component: (
+          <Grid container alignItems='flex-end' spacing={2} key={key}>
+            <Grid item xs={6}><Typography>{ formatString(key) }</Typography></Grid>
+            <Grid item xs={6}><StringInput objectKey={key} data={data}/></Grid>
+          </Grid>
+        )
+      },
+      {
+        dataType: 'long',
+        score: ( value === 'long' ? 60 : 10),
+        component: (
+          <Grid container alignItems='flex-end' spacing={2} key={key}>
+            <Grid item xs={6}><Typography>{ formatString(key) }</Typography></Grid>
+            <Grid item xs={6}><LongInput objectKey={key} data={data}/></Grid>
+          </Grid>
+        )
+      },
+      {
+        dataType: 'object',
+        score: ( typeof(value) === 'object' ? 60 : 0),
+        component: (<ObjectInput objectKey={key} value={value} data={data}></ObjectInput>)
+      }
```
This segment being hard-coded seems strange to me, given the lengths you went through to make everything else not hard-coded. Could we adopt the ComponentManager approach to generating this list?

nela-jankechova replied:
I will implement that today.

OrdiNeu commented on modules/data-entry/src/main/frontend/src/questionnaire/NewQuestionnaireDialog.jsx:
```diff
@@ -0,0 +1,144 @@
+        .finally(() => {setFetching(false)});
+      setFetching(true);
+    }
+  }
```
Should be an extra newline between functions for readability
```suggestion
  }

```

OrdiNeu commented on modules/data-entry/src/main/frontend/src/questionnaire/Questionnaire.jsx:
```diff
@@ -77,57 +109,206 @@ let Questionnaire = (props) => {
+  }
+  return (
+    <Card>
+      <CardHeader title={data['title'] || ''}
```
Sections might benefit from a default title, since it's a bit hard to see them at a glance. Wdyt @marta-?

nela-jankechova replied:
While I was writing the code, I used the UUID as a placeholder when there was no title. Maybe that could work?

nela-jankechova committed:
Update modules/data-entry/src/main/frontend/src/questionnaire/StringInput.jsx

Co-authored-by: OrdiNeu <OrdinaryNeumann@gmail.com>

nela-jankechova committed:
Update modules/data-entry/src/main/frontend/src/questionnaire/Fields.jsx

Co-authored-by: OrdiNeu <OrdinaryNeumann@gmail.com>

nela-jankechova committed:
Update modules/data-entry/src/main/frontend/src/questionnaire/NewQuestionnaireDialog.jsx

Co-authored-by: OrdiNeu <OrdinaryNeumann@gmail.com>

nela-jankechova committed:
Update modules/data-entry/src/main/frontend/src/questionnaire/BooleanInput.jsx

Co-authored-by: OrdiNeu <OrdinaryNeumann@gmail.com>

nela-jankechova committed:
Update modules/data-entry/src/main/frontend/src/questionnaire/DeleteDialog.jsx

Co-authored-by: OrdiNeu <OrdinaryNeumann@gmail.com>

nela-jankechova committed:
Update modules/data-entry/src/main/frontend/src/questionnaire/Question.json

Co-authored-by: OrdiNeu <OrdinaryNeumann@gmail.com>

nela-jankechova committed:
Update modules/data-entry/src/main/frontend/src/questionnaire/ObjectInput.jsx

Co-authored-by: OrdiNeu <OrdinaryNeumann@gmail.com>

nela-jankechova committed:
Update Section.json

nela-jankechova committed:
LFS-172: Refactor to use ComponentManager Method

OrdiNeu commented on modules/data-entry/src/main/frontend/src/dataHomepage/Questionnaires.jsx:
```diff
@@ -43,8 +56,36 @@ export default function Questionnaires(props) {
+    <React.Fragment>
+      <Card>
+       <CardHeader
+        color={"warning"/* Does nothing */}
```
```suggestion
```

OrdiNeu commented on modules/data-entry/src/main/frontend/src/dataHomepage/Questionnaires.jsx:
```diff
@@ -43,8 +56,36 @@ export default function Questionnaires(props) {
   );
 }
+
+export default withStyles(QuestionnaireStyle)(Questionnaires);
```
```suggestion
export default withStyles(QuestionnaireStyle)(Questionnaires);

```

OrdiNeu commented on modules/data-entry/src/main/frontend/src/questionnaire/AnswerOptions.jsx:
```diff
@@ -0,0 +1,145 @@
+
+    if (!newValue.includes(tempValue)) {
+      console.log("newValue", newValue);
+      console.log("tempValue", tempValue);
```
```suggestion
```

OrdiNeu commented on modules/data-entry/src/main/frontend/src/questionnaire/AnswerOptions.jsx:
```diff
@@ -0,0 +1,145 @@
+      { newUuid.map((value, index) =>
+        <Grid item xs={6} key={index}>
+          <input type='hidden' name={`${props.data['@path']}/${newValue[index]}/jcr:primaryType`} value={'lfs:AnswerOption'} />
+          {/* <input type='hidden' name={`${props.data['@path']}/${newValue[index]}/jcr:uuid`} value={value} /> */}
```
```suggestion
```

OrdiNeu commented on modules/data-entry/src/main/frontend/src/questionnaire/DeleteDialog.jsx:
```diff
@@ -0,0 +1,126 @@
+  let loginToSave = () => {
+    const width = 600;
+    const height = 800;
+    const top = window.top.outerHeight / 2 + window.top.screenY - ( height / 2);
```
```suggestion
    const top = window.top.outerHeight / 2 + window.top.screenY - (height / 2);
```

OrdiNeu commented on modules/data-entry/src/main/frontend/src/questionnaire/DeleteDialog.jsx:
```diff
@@ -0,0 +1,126 @@
+    const width = 600;
+    const height = 800;
+    const top = window.top.outerHeight / 2 + window.top.screenY - ( height / 2);
+    const left = window.top.outerWidth / 2 + window.top.screenX - ( width / 2);
```
```suggestion
    const left = window.top.outerWidth / 2 + window.top.screenX - (width / 2);
```

OrdiNeu commented on modules/data-entry/src/main/frontend/src/questionnaire/DeleteDialog.jsx:
```diff
@@ -0,0 +1,126 @@
+      <Dialog id="deleteDialog" open={openDeleteDialog} onClose={() => { setOpenDeleteDialog(false); }}>
+        <form action={props.data["@path"]} onSubmit={deleteData} method="DELETE" key={props.id}>
+          <DialogTitle>
+            <Typography>{props.type.includes("Question") ? "Confirm question deletion" : "Confirm Section Deletion"}</Typography>
```
```suggestion
            <Typography>{props.type.includes("Question") ? "Confirm Question Deletion" : "Confirm Section Deletion"}</Typography>
```

OrdiNeu commented on modules/data-entry/src/main/frontend/src/questionnaire/DeleteQuestionnaireDialog.jsx:
```diff
@@ -0,0 +1,121 @@
+import QuestionnaireStyle from "./QuestionnaireStyle.jsx";
+
+let DeleteQuestionnaireDialog = (props) => {
+    let [ forms, setForms] = useState(0);
```
```suggestion
    let [ forms, setForms ] = useState(0);
```

OrdiNeu commented on modules/data-entry/src/main/frontend/src/questionnaire/DeleteQuestionnaireDialog.jsx:
```diff
@@ -0,0 +1,121 @@
+      const width = 600;
+      const height = 800;
+      const top = window.top.outerHeight / 2 + window.top.screenY - ( height / 2);
+      const left = window.top.outerWidth / 2 + window.top.screenX - ( width / 2);
```
```suggestion
      const top = window.top.outerHeight / 2 + window.top.screenY - (height / 2);
      const left = window.top.outerWidth / 2 + window.top.screenX - (width / 2);
```

sashaandjic commented:
When creating a new section, name and title get concatenated. Step to reproduce:
- create new section
![image](https://user-images.githubusercontent.com/23243244/87963589-f13da980-ca86-11ea-8d16-878432af12e2.png)
- save>refresh>observe 
![image](https://user-images.githubusercontent.com/23243244/87963649-103c3b80-ca87-11ea-8d54-cbab38af275c.png)
- go to edit mode and notice that in field name is no longer available (name and title are stored in the title field comma separated)
![image](https://user-images.githubusercontent.com/23243244/87963780-44176100-ca87-11ea-8c4f-4cd2011f0b04.png)

sashaandjic commented:
I believe there is a regression that was recently introduced. On my older instance of LFS-172 build I get dropdown filed for data type
![image](https://user-images.githubusercontent.com/23243244/87968131-efc3af80-ca8d-11ea-9143-847920c20d30.png)
On my new instance that I just built I get this:
![image](https://user-images.githubusercontent.com/23243244/87968238-17b31300-ca8e-11ea-9f3d-ddf06992a8c5.png)

OrdiNeu commented on modules/data-entry/src/main/frontend/src/questionnaire/EditDialog.jsx:
```diff
@@ -0,0 +1,191 @@
+//
```
There's a lot of identical code between the three kinds of Dialog (EditDialog, DeleteDialog, DeleteQuestionnaireDialog). Can we refactor the common components between them into one class, and pass the parts that differ as props?

OrdiNeu commented on modules/data-entry/src/main/frontend/src/dataHomepage/Questionnaires.jsx:
```diff
@@ -16,17 +16,30 @@
+function Questionnaires(props) {
+  let [ openDialog, setOpenDialog] = useState(false);
+  let [ deleteData, setDeleteData ] = useState(false);
+  const { match, classes } = props;
```
```suggestion
  const { classes } = props;
```

OrdiNeu commented on modules/data-entry/src/main/frontend/src/questionnaire/Fields.jsx:
```diff
@@ -0,0 +1,63 @@
+ * @returns a React component that renders the question
+ */
+
+let displayField = (key, value) => {
+  // This variable must start with an upper case letter so that React treats it as a component
+  const FieldDisplay = QuestionComponentManager.getQuestionComponent(key);
+  return (
+    <Grid item key={key}>
+      <FieldDisplay
+        objectKey={key}
+        data={data}
+        />
+    </Grid>
+  );
+};
```
Indentation

OrdiNeu commented on modules/data-entry/src/main/frontend/src/questionnaire/Fields.jsx:
```diff
@@ -0,0 +1,63 @@
+};
+
+  return Object.entries(JSON).map(([key, value]) => (
+    displayField(key, value)));
```
```suggestion
    displayField(key, value))
    );
```

OrdiNeu commented on modules/data-entry/src/main/frontend/src/questionnaire/ObjectInput.jsx:
```diff
@@ -0,0 +1,84 @@
+
+let ObjectInput = (props) => {
+  let { objectKey, value, data, definition } = props;
+  let [ selectedValue, setSelectedValue] = useState(data[objectKey] || '');
```
```suggestion
  let [ selectedValue, setSelectedValue ] = useState(data[objectKey] || '');
```

OrdiNeu commented on modules/data-entry/src/main/frontend/src/questionnaire/ObjectInput.jsx:
```diff
@@ -0,0 +1,84 @@
+// Object Input field used by Edit dialog component
+
+let ObjectInput = (props) => {
+  let { objectKey, value, data, definition } = props;
```
```suggestion
  let { objectKey, value, data } = props;
```

OrdiNeu commented on modules/data-entry/src/main/frontend/src/questionnaire/Fields.jsx:
```diff
@@ -0,0 +1,63 @@
+
+import React from 'react';
+import PropTypes from 'prop-types';
+import { Grid, Typography, withStyles } from "@material-ui/core";
```
```suggestion
import { Grid, withStyles } from "@material-ui/core";
```

OrdiNeu commented on modules/data-entry/src/main/frontend/src/questionnaire/ObjectInput.jsx:
```diff
@@ -0,0 +1,84 @@
+
+QuestionComponentManager.registerQuestionComponent((definition) => {
+  if (typeof(definition) === 'object') {
+    return [StyledObjectInput, 50];
```
I think something's up with this; the dataType isn't properly being picked up as an object input:
![image](https://user-images.githubusercontent.com/4656440/88067526-18ef4900-cb3d-11ea-9413-6d0e5915f4f6.png)

OrdiNeu reviewed (changes requested):
Can we move the Questionnaire editing code to its own folder? The questionnaire/ folder is getting cramped with a bunch of unrelated bits of code.

nela-jankechova committed:
LFS-172: Auto-update questionnaire

nela-jankechova committed:
Update modules/data-entry/src/main/frontend/src/questionnaire/Fields.jsx

Co-authored-by: OrdiNeu <OrdinaryNeumann@gmail.com>

nela-jankechova committed:
Update modules/data-entry/src/main/frontend/src/dataHomepage/Questionnaires.jsx

Co-authored-by: OrdiNeu <OrdinaryNeumann@gmail.com>

nela-jankechova committed:
Update modules/data-entry/src/main/frontend/src/questionnaire/AnswerOptions.jsx

Co-authored-by: OrdiNeu <OrdinaryNeumann@gmail.com>

nela-jankechova committed:
Update modules/data-entry/src/main/frontend/src/questionnaire/AnswerOptions.jsx

Co-authored-by: OrdiNeu <OrdinaryNeumann@gmail.com>

nela-jankechova committed:
Update modules/data-entry/src/main/frontend/src/questionnaire/ObjectInput.jsx

Co-authored-by: OrdiNeu <OrdinaryNeumann@gmail.com>

nela-jankechova committed:
Update modules/data-entry/src/main/frontend/src/questionnaire/DeleteDialog.jsx

Co-authored-by: OrdiNeu <OrdinaryNeumann@gmail.com>

nela-jankechova committed:
Update modules/data-entry/src/main/frontend/src/questionnaire/DeleteDialog.jsx

Co-authored-by: OrdiNeu <OrdinaryNeumann@gmail.com>

nela-jankechova committed:
Update modules/data-entry/src/main/frontend/src/questionnaire/ObjectInput.jsx

Co-authored-by: OrdiNeu <OrdinaryNeumann@gmail.com>

nela-jankechova committed:
Update modules/data-entry/src/main/frontend/src/questionnaire/DeleteDialog.jsx

Co-authored-by: OrdiNeu <OrdinaryNeumann@gmail.com>

nela-jankechova committed:
Update modules/data-entry/src/main/frontend/src/questionnaire/DeleteQuestionnaireDialog.jsx

Co-authored-by: OrdiNeu <OrdinaryNeumann@gmail.com>

nela-jankechova committed:
Update modules/data-entry/src/main/frontend/src/questionnaire/DeleteQuestionnaireDialog.jsx

Co-authored-by: OrdiNeu <OrdinaryNeumann@gmail.com>

nela-jankechova committed:
LFS-172: Fix Object Bug & Move into new Folder

sashaandjic commented:
I created a new question for the Chemotherapy Data:

![image](https://user-images.githubusercontent.com/23243244/89826283-05277900-db24-11ea-8f72-58dc42560866.png)

Then I wanted to add a new section to the same questionnaire. The value of the question title previously entered got copied into the section name ( and vice versa - section name would get copied into the question title if I reversed the oreder of creation).

![image](https://user-images.githubusercontent.com/23243244/89825876-4703ef80-db23-11ea-9c44-e7cee54fc5df.png)


![image](https://user-images.githubusercontent.com/23243244/89825928-5edb7380-db23-11ea-87a0-f0a10643e538.png)

OrdiNeu committed:
LFS-172: As an admin, I can import, visualize and design questionnaires
Allow the enter key to create questionnaires when the new questionnaire
dialog is open.

OrdiNeu committed:
LFS-172: As an admin, I can import, visualize and design questionnaires
Fix a broken redirect

OrdiNeu committed:
LFS-172: As an admin, I can import, visualize and design questionnaires
Fix up a broken redirect

OrdiNeu committed:
LFS-172: As an admin, I can import, visualize and design questionnaires
Fix a bug where sections would have question options.

OrdiNeu committed:
LFS-172: As an admin, I can import, visualize and design questionnaires
Fix the boolean input not working at all.

OrdiNeu committed:
LFS-172: As an admin, I can import, visualize and design questionnaires
Code cleanup

OrdiNeu committed:
LFS-172: As an admin, I can import, visualize and design questionnaires
Further code cleanup

OrdiNeu committed:
LFS-172: As an admin, I can import, visualize and design questionnaires
Improve error reporting

OrdiNeu committed:
LFS-172: As an admin, I can import, visualize and design questionnaires
Force reload of the questionnaires page when the user deletes a
questionnaire.

OrdiNeu committed:
LFS-172: As an admin, I can import, visualize and design questionnaires
Prevent the title from reappearing between new questions.

OrdiNeu committed:
LFS-172: As an admin, I can import, visualize and design questionnaires
Fix a bug where both the name and title were being appended to the title
of a Section

sashaandjic commented:
I was able to design a new questionnaire, however I was not able to get the constraints work ( e.g. with min answers is 1& max 2: I can add as many as I want) . Also, not sure how to create a radio bttn or design a field such as Tumor's malignancy filed.

sashaandjic commented:
This definition renders field as read only
![image](https://user-images.githubusercontent.com/23243244/91496234-c31f6680-e889-11ea-9aff-4de05ac54c67.png)

sashaandjic commented:
This setting allows only 2 entries ( should be 3)
![image](https://user-images.githubusercontent.com/23243244/91496557-52c51500-e88a-11ea-9353-b265f85041f2.png)

sashaandjic commented:
The Action column should be pinned to the right.
![image](https://user-images.githubusercontent.com/23243244/91529022-888feb00-e8d6-11ea-87ac-a85f5455ee02.png)

sashaandjic commented:
Two outstanding problems are reported in: LFS-540 ; LFS-542

OrdiNeu committed:
LFS-172: As an admin, I can import, visualize and design questionnaires
Move the actions column to the end

OrdiNeu commented:
> The Action column should be pinned to the right.
> ![image](https://user-images.githubusercontent.com/23243244/91529022-888feb00-e8d6-11ea-87ac-a85f5455ee02.png)

@sashaandjic This should now be fixed. The other bug will be fixed in #271

sashaandjic commented:
@OrdiNeu did you include changes from #271 in this PR. I do not see them. I only see this one being fixed
![image](https://user-images.githubusercontent.com/23243244/92176669-f4a9ac00-ee0c-11ea-8565-0d6ba7679ed5.png)

sashaandjic commented:
![image](https://user-images.githubusercontent.com/23243244/92176698-055a2200-ee0d-11ea-81ce-177d02aa10cd.png)
![image](https://user-images.githubusercontent.com/23243244/92176714-0be89980-ee0d-11ea-8a23-51c5ccff9999.png)

sashaandjic commented:
Two more problems found and reported in separate tasks :
- LFS-544 - unable to delete questions placed withing section
- LFS-545 - unable to delete section

OrdiNeu commented:
Max Answers and the actions on the right side have now been fixed. Ready for re-test.

sashaandjic commented:
Max Answers and the actions on the right side are fixed,
I am unable to remove Answer Options from questions once they are defined.
![image](https://user-images.githubusercontent.com/23243244/92644417-5062b200-f2b1-11ea-80f8-6fac28718956.png)
![image](https://user-images.githubusercontent.com/23243244/92644388-42149600-f2b1-11ea-9d8e-00e9bc74e2a5.png)
I deleted all of them:
![image](https://user-images.githubusercontent.com/23243244/92644492-72f4cb00-f2b1-11ea-996e-6a09c1ef6890.png)
Clicked on SAVE bttn, however they are all still there.

sashaandjic commented:
I would also like to be able to move questions within questionnaire ( probably out of scope).

sashaandjic commented:
How can I create a conditional section?

sashaandjic commented:
Min and Max value are not enforced. Note that the last character is cut of the Description field once you go to edit mode. 
![image](https://user-images.githubusercontent.com/23243244/92645972-b819fc80-f2b3-11ea-8a25-2d1a72a144ba.png)
![image](https://user-images.githubusercontent.com/23243244/92645988-bf410a80-f2b3-11ea-8234-fd4ed5cf4286.png)

OrdiNeu committed:
LFS-172: As an admin, I can import, visualize and design questionnaires
Remove old AnswerOptions

OrdiNeu committed:
LFS-172: As an admin, I can import, visualize and design questionnaires
Fixing the logic when answer options are deleted or inserted into a new
form

OrdiNeu committed:
LFS-172: As an admin, I can import, visualize and design questionnaires
Fix the minValue/maxValue variables

OrdiNeu committed:
LFS-172: As an admin, I can import, visualize and design questionnaires
Add multiline to every input field to allow for long descriptions

sashaandjic commented:
I am still unable to delete questions and sections. 
When I specify that Min Answers is > 0, I would expect the field to be required, however I can leave such field blank.

OrdiNeu committed:
LFS-172: As an admin, I can import, visualize and design questionnaires
Fix bug where deleting sections would not work.

OrdiNeu committed:
LFS-172: As an admin, I can import, visualize and design questionnaires
Add better error message handling

OrdiNeu committed:
LFS-172: As an admin, I can import, visualize and design questionnaires
Recursively delete answers when a question is deleted

OrdiNeu committed:
LFS-172: As an admin, I can import, visualize and design questionnaires
Make the error message a bit more robust

sashaandjic commented:
Section deletion now works. I can delete sections from questionnaires with active forms and with data entered in them. 
---------
An error msg warsn me that I cannot delete questionnaire as I have an active form of that type. There is also message at the bottom of the screen.
![image](https://user-images.githubusercontent.com/23243244/93646230-d4184f00-f9d3-11ea-8359-f36f890f39b6.png)
Once I close the dialog, the message in red fonts at the bottom of the screen is still there. 
![image](https://user-images.githubusercontent.com/23243244/93646401-248fac80-f9d4-11ea-9f53-42d6a66d0172.png)

OrdiNeu committed:
LFS-172: As an admin, I can import, visualize and design questionnaires
Make recursive deletions clear children first
Make sections delete-recursiveable

OrdiNeu committed:
LFS-172: As an admin, I can import, visualize and design questionnaires
Use recursive deletion of questionnaires

OrdiNeu committed:
LFS_172: As an admin, I can import, visualize and design questionnaires
Apply the recursive deletion to every type of node. Fix the
questionnaire warning.

OrdiNeu commented:
> This is good, but it makes the "warning" displayed to the user inaccurate. Do we care?
I've made the warning more generic in `cf29fe` :heavy_check_mark: .
> Should we make this more generic and let it handle deletes for any kind of resource?
I've switched this to allow it to handle deletion of any of our resources (lfs/Resource). :heavy_check_mark: 
Not sure if we want this functionality across any `nt:` or `sling:` nodes yet.

marta- reviewed (approved):
Ok for now. Some adjustments to be implemented as separate tasks.

sashaandjic commented:
Problems found so far:
- Predifined questionsrries: cannot delete sections (with or without data), cannot delete questionarries
![image](https://user-images.githubusercontent.com/23243244/94181484-8edd9c80-fe6d-11ea-86e5-accd00d1fc3b.png)
![image](https://user-images.githubusercontent.com/23243244/94181501-97ce6e00-fe6d-11ea-8a17-d017e336d595.png)


- when creating a new question with Answer Options, the Answer Options are not saved
![image](https://user-images.githubusercontent.com/23243244/94181319-50e07880-fe6d-11ea-8764-d66083517d4e.png)
I cliked on SAVE and opened the question in edit mode
![image](https://user-images.githubusercontent.com/23243244/94181396-6b1a5680-fe6d-11ea-8c1d-cbd0e530aae9.png)
